### PR TITLE
[USH-3239] reverse index query optimization

### PR DESCRIPTION
**Pull request by snopoke**
Jira: https://dimagi-dev.atlassian.net/browse/USH-3239
## Technical Summary
This PR contains the platform changes necessary to optimize the reverse index predicate: `[index/parent = current()/@case_id]`.

This predicate is already partially optimized to do a bulk lookup for each value of `current()/@case_id`. What this PR does is fully optimize the lookups so that a single query is performed to find all the index cases for all the cases in the current scope.

This data is then cached and used for later evaluation.

### Safety story
The safety of this should be validated by unit tests.

### Automated test coverage
This XPath evaluation code is well covered by tests and I have added tests explicitly for this optimization in the Formplayer PR. 

### QA Plan
I will put this on staging for final testing just to be safe but do not intend to open a QA ticket for it.

### Special deploy instructions

- [x] This PR can be deployed after merge with no further considerations.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations.

### Review

- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change.

### CODE REVIEW

**Review by shubham1g5 — APPROVED** [review]
looks great!

**Comment by ctsims** [comment]
@snopoke Awesome change, at this is at just the right place in the engine I think.

I poked around in the engine a bit to refresh my memory, and there are a few edge cases I think would also be helpful to add to the test set.

The first would be a scenario where the basis for the `index/parent` transform **isn't** the the same as the `current()` base that's populated from the QueryContext. That's the most likely place for the engine to get confused. It would be a pattern like:

`instance('casedb')/casedb/case[other_optimizable_filter][index/parent = current()/...]`

the other scenario is making sure that the structure chains, ie: Where `[index/parent = current()/@case_id]` is nested in another predicate. That's the main goal of Model Set transforms, but it's also just a place where things are the most likely to just crash outright if anything's off.

File: src/main/java/org/commcare/modern/engine/cases/ReverseCaseIndexQuerySetTransform.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import org.commcare.cases.query.queryset.QuerySetLookup;
+import org.commcare.cases.query.queryset.QuerySetTransform;
+import org.commcare.modern.util.PerformanceTuningUtil;
+import org.javarosa.core.model.instance.TreeReference;
+import org.javarosa.core.model.trace.EvaluationTrace;
+
+import java.util.Set;
+
+/**
+ * Created by skelly on 2023/05/17.
```
Comment (ctsims) [inline src/main/java/org/commcare/modern/engine/cases/ReverseCaseIndexQuerySetTransform.java:18]: Javadoc outline for this would be really helpful.

**Review by ctsims — APPROVED** [review]
This looks great, thanks for taking the time to find the moving pieces to wire it in.

It would be good to update the docs at 
https://github.com/dimagi/commcare-core/wiki/Performance:-Query-Planner#structure

to reflect the new availability of this model set pattern.

**Comment by ctsims** [comment]
Another quick note that I checked the trace instrumentation on [nested contexts](https://github.com/dimagi/commcare-android/pull/2674/files#diff-f326941b39367eb33a0124fd7fd9cb4778454c345d0cf77035d6b393ca115f3aR136) and it looks like it's working correctly as long as the queryset lookup volumes pass the bulk threshold, which is awesome to see. 

Also, when reviewing that test case it looks like even though the Single Loads might have a performance regression, bulk loads don't appear to be affected, the original query plan has the ~same bulk load time (10ms longer in the pre-PR query path).
